[14.0] [IMP] Gather all the UserError as one when validating payments in payment order

simgonzalez opened the pull request:
PR top add the feature of showing all the validation user errors at once and for the user not to have to launch 10 times the process to get errors one by one.

pedrobaeza commented:
Please follow commit message guidelines:

https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#71commit-message

- No version on the commit message, but on the PR title.
- `[FIX]` tag
- Name of the module
- Short explanation of what in the first line
- Long explanation, whys and so on on following lines

pedrobaeza commented on account_payment_order/models/account_payment_order.py:
```diff
@@ -344,6 +348,13 @@ def draft2open(self):
+            # Raise errors that happened on the validation process
+            if payline_err_text:
+                raise UserError(
+                    _("A bunch of validation didn't passed\n")
```
```suggestion
                    _("There's at least one validation error:\n")
```

simgonzalez replied:
Is there a way to check pre-commit stuff before actually the PR check ?

pedrobaeza replied:
You can install pre-commit locally. Having pipx installed (`pip install pipx`), do `pipx install pre-commit`. Then, being on your repo directory, do `pre-commit install -f` for installing a hook. Now, each `git commit` on that repo, will launch first the pre-commit stuff.

pedrobaeza reviewed (commented):
And check pre-commit stuff

pedrobaeza commented on account_payment_order/models/account_payment_order.py:
```diff
@@ -333,17 +336,26 @@ def draft2open(self):
+                        group_paylines[hashcode]["total"] += payline.amount_currency
+                    else:
+                        group_paylines[hashcode] = {
+                            "paylines": payline,
+                            "total": payline.amount_currency,
+                        }
                 else:
-                    # Use line ID as hascode, which actually means no grouping
-                    hashcode = payline.id
-                if hashcode in group_paylines:
-                    group_paylines[hashcode]["paylines"] += payline
-                    group_paylines[hashcode]["total"] += payline.amount_currency
-                else:
-                    group_paylines[hashcode] = {
+                    group_paylines[payline.id] = {
                         "paylines": payline,
                         "total": payline.amount_currency,
                     }
```
Don't repeat code:

```suggestion
                hashcode = payline.payment_line_hashcode() if order.payment_mode_id.group_lines else payline.id
                if hashcode in group_paylines:
                    group_paylines[hashcode]["paylines"] += payline
                    group_paylines[hashcode]["total"] += payline.amount_currency
                else:
                    group_paylines[hashcode] = {
                        "paylines": payline,
                        "total": payline.amount_currency,
                    }
```

Commit:
[IMP] account_payment_order: Concatenate all UserError messages
Gather all the errors to only show one with all the problem on the current debit order. This will avoid for the users to have to make multiple time the same process.

pedrobaeza commented on account_payment_order/models/account_payment_order.py:
```diff
@@ -331,11 +334,11 @@ def draft2open(self):
-                else:
-                    # Use line ID as hascode, which actually means no grouping
-                    hashcode = payline.id
+                hashcode = (
```
Now that the diff is clearer, I'm seeing that previous code was OK. Why did you initially change it? I think it's better to leave it as it was for reducing diff. Sorry for changing my mind.

simgonzalez replied:
I know but the pre-commit didn't let me commit as it was because of complexity factor that was too high. This is the reason why i modified this part

pedrobaeza replied:
OK then.

simgonzalez commented:
@pedrobaeza Do you know to who i should ask a second review ? Before this PR automatically closes :D 
Thanks ^^

pedrobaeza commented:
You can fw-port it to upper versions:

/ocabot merge patch
